vms/platformvm/state/stakers_model_storage_test.go

```diff
@@ -0,0 +1,966 @@
+	return true
+}
+
+func (cmd *rebuildStateCommand) PostCondition(cmdState commands.State, res commands.Result) *gopter.PropResult {
+	if cmd.err != nil {
+		cmd.err = nil // reset for next runs
+		return &gopter.PropResult{Status: gopter.PropFalse}
+	}
+
+	if !checkSystemAndModelContent(cmdState, res) {
+		return &gopter.PropResult{Status: gopter.PropFalse}
+	}
+
+	if !checkValidatorSetContent(res) {
+		return &gopter.PropResult{Status: gopter.PropFalse}
+	}
+
+	return &gopter.PropResult{Status: gopter.PropTrue}
+}
+
+func (*rebuildStateCommand) String() string {
+	return "\nrebuildStateCommand"
+}
+
+// a trick to force command regeneration at each sampling.
+// gen.Const would not allow it
+var genRebuildStateCommand = gen.IntRange(1, 2).Map(
+	func(int) commands.Command {
+		return &rebuildStateCommand{}
+	},
+)
+
+func checkSystemAndModelContent(cmdState commands.State, res commands.Result) bool {
+	model := cmdState.(*stakersStorageModel)
+	sys := res.(*sysUnderTest)
+
+	// top view content must always match model content
+	topDiff := sys.getTopChainState()
+
+	modelIt, err := model.GetCurrentStakerIterator()
+	if err != nil {
+		return false
+	}
+	sysIt, err := topDiff.GetCurrentStakerIterator()
+	if err != nil {
+		return false
+	}
+
+	modelStakers := make([]*Staker, 0)
+	for modelIt.Next() {
+		modelStakers = append(modelStakers, modelIt.Value())
+	}
+	modelIt.Release()
+
+	sysStakers := make([]*Staker, 0)
+	for sysIt.Next() {
+		sysStakers = append(sysStakers, sysIt.Value())
+	}
+	sysIt.Release()
+
+	if len(modelStakers) != len(sysStakers) {
+		return false
+	}
+
+	for idx, modelStaker := range modelStakers {
+		sysStaker := sysStakers[idx]
+		if modelStaker == nil || sysStaker == nil || !reflect.DeepEqual(modelStaker, sysStaker) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func checkValidatorSetContent(res commands.Result) bool {
+	sys := res.(*sysUnderTest)
+	valSet := sys.baseState.(*state).cfg.Validators
+
+	sysIt, err := sys.baseState.GetCurrentStakerIterator()
+	if err != nil {
+		return false
+	}
+
+	// valContent subnetID -> nodeID -> aggregate weight (validator's own weight + delegators' weight)
+	valContent := make(map[ids.ID]map[ids.NodeID]uint64)
+	for sysIt.Next() {
+		val := sysIt.Value()
+		if val.SubnetID != constants.PrimaryNetworkID {
+			continue
+		}
+		nodes, found := valContent[val.SubnetID]
+		if !found {
+			nodes = make(map[ids.NodeID]uint64)
+			valContent[val.SubnetID] = nodes
+		}
+		nodes[val.NodeID] += val.Weight
+	}
+	sysIt.Release()
+
+	for subnetID, nodes := range valContent {
+		vals, found := valSet.Get(subnetID)
+		if !found {
+			return false
+		}
+		for nodeID, weight := range nodes {
+			valWeight := vals.GetWeight(nodeID)
+			if weight != valWeight {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+type sysUnderTest struct {
+	diffBlkIDSeed uint64
+	baseDB        database.Database
+	baseState     State
+	sortedDiffIDs []ids.ID
```

Comment:
queue of diff id in the order they are created. Recall that diffs are processed in FIFO order: created on top, consumed at the bottom (when applied to baseState)